### PR TITLE
New mod content installer

### LINKED ISSUE #7412: Video installation
`ExtractFilesFromCD` needs an optional flag for `~movies1.mix, ~movies2.mix` and we could think about if our mirror network is strong enough to handle the downloads of those files (around 400 MB each).

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8186860-video-installation?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #10556: Error messages during mod installation are not truncated to fit the dialog
screen shot 2016-01-16 at 11 40 53

We may want to implement new options for truncating strings so that we can specify that the start of the path should be chopped off (which is at an arbitrary position inside the string).

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30064752-error-messages-during-mod-installation-are-not-truncated-to-fit-the-dialog?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #11344: OpenRA.Utility.exe ra --extract does not work as described on linux
OpenRA.Utility.exe ra --extract does not work as described here: https://github.com/OpenRA/OpenRA/wiki/Game-Content on a debian GNU/linux system. Steps to reproduce:
1. put MAIN.MIX from "Red Alert Allied" CD into folder $Home/.openra/Content/ra
2. open terminal and do:
```
cd /usr/share/games/openra
mono --debug ./OpenRA.Utility.exe ra --extract movies1.mix
```

Expected result: file movies1.mix in folder /usr/share/games/openra.

Actual result: The following error message
```
Error: Utility application crashed. See utility.log for details

Unhandled Exception:
System.IO.FileNotFoundException: File not found: movies1.mix
File name: 'movies1.mix'
  at OpenRA.FileSystem.FileSystem.Open (System.String filename) <0x41312470 + 0x000af> in <filename unknown>:0 
  at OpenRA.Mods.Common.UtilityCommands.ExtractFilesCommand.Run (OpenRA.ModData modData, System.String[] args) <0x413373f0 + 0x0009d> in <filename unknown>:0 
  at OpenRA.Utility.Program.Main (System.String[] args) <0x412d4040 + 0x005f0> in <filename unknown>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.IO.FileNotFoundException: File not found: movies1.mix
File name: 'movies1.mix'
  at OpenRA.FileSystem.FileSystem.Open (System.String filename) <0x41312470 + 0x000af> in <filename unknown>:0 
  at OpenRA.Mods.Common.UtilityCommands.ExtractFilesCommand.Run (OpenRA.ModData modData, System.String[] args) <0x413373f0 + 0x0009d> in <filename unknown>:0 
  at OpenRA.Utility.Program.Main (System.String[] args) <0x412d4040 + 0x005f0> in <filename unknown>:0 
```

In fact, when in the same folder as MAIN.MIX and typing
```
mono /usr/share/games/openra/OpenRA.Utility.exe ra --extract MAIN.MIX
```
it says
```
MAIN.MIX saved.
```
but nothing else happens.

Am I missing something concerning usage of the utility? Is there a technical reason not to just have the program take two arguments, namely file name of the archive (MAIN.MIX) and file name of the desired file to extract?

### PR DESCRIPTION
**Pull request by pchote**
The original content installer has been hacked together over the years, and had a whole bunch of limitations.  This PR completely rewrites it, bringing the following benefits:
- Now understands the difference between different content sources, making it easy to add new ones in the future (Origin / cncnet / GOG etc, as well as translated versions).
- Can now install disk-specific assets (FMVs, expansion music)
- Improved discovery for music installation
- Includes mod-specific description text, making it clear that our asset downloads come from the freeware TD/RA/TS releases
- Removes the modchooser dependency on game-specific package parsers (Installshield, Mix, etc)

There are currently two minor regressions which I would like to leave to followup PRs (the PR is big enough as it is, and these are separate to the main changes):
- We lose TFD support until somebody who has access to the disk can set up the new metadata for it
- The music installation prompt is not shown for players who have the hacked scores.mix from the original assets download.  A future PR will add metadata that tells the game to delete that file as a migration step

Closes #7412, Closes #11344, Closes #10556
Related to #9479, #5970, #8993, #4361, #9082

screen shot 2016-05-29 at 19 55 37
screen shot 2016-05-29 at 19 56 03
screen shot 2016-05-29 at 19 56 07
screen shot 2016-05-29 at 19 56 32

### CODE REVIEW

**Comment by jaZzKCS** [comment]
Very awesome work.
:+1:

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 1]
```diff
@@ -748,6 +748,12 @@
     <Compile Include="Traits\Radar\JamsRadar.cs" />
     <Compile Include="AI\DummyAI.cs" />
     <Compile Include="UtilityCommands\ListInstallShieldContentsCommand.cs" />
+    <Compile Include="Widgets\Logic\Installation\ModContentLogic.cs" />
+    <Compile Include="Widgets\Logic\Installation\DownloadPackageLogic.cs" />
+    <Compile Include="Widgets\Logic\Installation\ModContentDiscTooltipLogic.cs" />
+    <Compile Include="Widgets\Logic\Installation\InstallFromDiscLogic.cs" />
+    <Compile Include="Widgets\Logic\Installation\ModContentPromptLogic.cs" />
```
Comment (Mailaender) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: These file additions are unrelated to this commit.

**Comment by pchote** [comment]
I think this should gracefully handle an empty mirror list and show a "Quick download is not available, please select Advanced install to copy assets from an original game disk" dialog.  This will avoid breakage if we decide/need to remove our online asset mirrors.

**Comment by pchote** [comment]
Updated:

* Fixed csproj additions
* Fixed the margins and improved the wording of the asset installation prompt:
screen shot 2016-05-31 at 22 52 05
* Added a graceful fallback if no mirrors are available:
screen shot 2016-05-31 at 22 50 49
* Added truncation for long status labels (fixes #10556)

**Comment by pchote** [comment]
This uses Assets/Content inconsistently through the UI.  We should standardize on one or the other.

My preference is probably Content.  Does anyone else have an opinion?

**Comment by Mailaender** [comment]
Feedback from social media doesn't indicate that it ever resulted in confusion or misunderstandings, so you are probably just fixing your personal urge to unify the wordings, but that is fine as well.

**Comment by pchote** [comment]
Updated.

File: OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+					return;
+				}
+
+				if (cancelled)
+				{
+					onError("Download cancelled");
+					return;
+				}
+
+				// Automatically extract
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs]: What if I don't want the download to be extracted, but just copied straight to Contents/mymod/? Is that an unreasonable use case? Afaik, we support .zip as archive for game files, and I doubt most people would go to the trouble of creating .mix files.

**Comment by obrakmann** [comment]
Code looks good so far. Haven't actually installed anything with it yet, will do that tomorrow.

File: OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs (hunk 2 above)
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs]: It's not afaik, I have successfully used zip files in place of mixes and I still plan to do that again.

**Comment by obrakmann** [comment]
Also, do you feel like writing a little blurb for https://github.com/OpenRA/OpenRA/wiki/Mod-manifest? :smile:

File: OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs (hunk 2 above)
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs]: I plan on doing exactly this – remember our earlier discussion about defining metadata to download byte ranges directly from a hosted ISO or installer exe?

That's another big job, so I'd prefer to leave this at feature parity with the current code for now.

File: OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+					Log.Write("install", "Extraction failed");
+
+					foreach (var f in extracted)
+					{
+						Log.Write("install", "Deleting " + f);
+						File.Delete(f);
+					}
+
+					onError("Invalid archive");
+				}
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs]: Please add a `finally` here that deletes the temp file.

**Comment by obrakmann** [comment]
Works very well, good job! :+1: once the temp file is properly cleaned up.

**Comment by pchote** [comment]
Added the temp file deletion.

**Comment by Mailaender** [comment]
Confirmation that the Dune 2000 video installation works on Windows from https://www.reddit.com/r/openra/comments/4n0rdw/how_to_play_dune2k_cutscenes/d44vfli

**Comment by Mailaender** [comment]
It currently crashes

```
Exception of type `System.Reflection.TargetInvocationException`: Exception has been thrown by the target of an invocation.
Inner Exception of type `System.ArgumentNullException`: Value cannot be null.
Parameter name: source
    at System.Linq.Enumerable.Select[TSource,TResult] (IEnumerable`1 source, System.Func`2 selector) [0x0006c] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/external/referencesource/System.Core/System/Linq/Enumerable.cs:44 
  at OpenRA.Mods.Common.Widgets.Logic.ModContentLogic.PopulateContentList () <0x40579520 + 0x004bb> in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.Logic.ModContentLogic..ctor (OpenRA.Widgets.Widget widget, System.String modId, System.Action onCancel) <0x40578a90 + 0x0088b> in <filename unknown>:0 
  at (wrapper managed-to-native) System.Reflection.MonoCMethod:InternalInvoke (System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00002] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:644 
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters) [0x00013] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:650 
  at System.Reflection.MonoCMethod.DoInvoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00089] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:635 
  at System.Reflection.MonoCMethod.Invoke (BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.Reflection/MonoMethod.cs:663 
  at System.Reflection.ConstructorInfo.Invoke (System.Object[] parameters) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.Reflection/ConstructorInfo.cs:62 
  at OpenRA.ObjectCreator.CreateUsingArgs (System.Reflection.ConstructorInfo ctor, System.Collections.Generic.Dictionary`2 args) [0x00062] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/ObjectCreator.cs:140 
  at OpenRA.ObjectCreator.CreateObject[T] (System.String className, System.Collections.Generic.Dictionary`2 args) [0x00063] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/ObjectCreator.cs:105 
  at OpenRA.Widgets.Widget+<PostInit>c__AnonStorey0.<>m__0 (System.String l) [0x00011] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:274 
  at System.Linq.Enumerable+WhereSelectArrayIterator`2[TSource,TResult].MoveNext () [0x0004d] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/external/referencesource/System.Core/System/Linq/Enumerable.cs:327 
  at System.Linq.Buffer`1[TElement]..ctor (IEnumerable`1 source) [0x00083] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/external/referencesource/System.Core/System/Linq/Enumerable.cs:2568 
  at System.Linq.Enumerable.ToArray[TSource] (IEnumerable`1 source) [0x00011] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/external/referencesource/System.Core/System/Linq/Enumerable.cs:830 
  at OpenRA.Widgets.Widget.PostInit (OpenRA.Widgets.WidgetArgs args) [0x0004c] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:275 
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, OpenRA.MiniYamlNode node) [0x001ea] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:75 
  at OpenRA.WidgetLoader.LoadWidget (OpenRA.Widgets.WidgetArgs args, OpenRA.Widgets.Widget parent, System.String w) [0x00032] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/WidgetLoader.cs:44 
  at OpenRA.Widgets.Ui.OpenWindow (System.String id, OpenRA.Widgets.WidgetArgs args) [0x00012] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:62 
  at OpenRA.Mods.Common.Widgets.Logic.ModBrowserLogic+<ModBrowserLogic>c__AnonStorey0.<>m__3 () <0x40578910 + 0x00107> in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.ButtonWidget.<ButtonWidget>m__A (MouseInput _) <0x40562990 + 0x0001f> in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.ButtonWidget.HandleMouseInput (MouseInput mi) <0x4055fed0 + 0x0023d> in <filename unknown>:0 
  at OpenRA.Widgets.Widget.HandleMouseInputOuter (MouseInput mi) [0x000ce] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:393 
  at OpenRA.Widgets.Ui.HandleInput (MouseInput mi) [0x0002c] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Widgets/Widget.cs:102 
  at OpenRA.DefaultInputHandler+<OnMouseInput>c__AnonStorey2.<>m__0 () [0x00006] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Input/InputHandler.cs:50 
  at OpenRA.Sync.CheckSyncUnchanged[T] (OpenRA.World world, System.Func`1 fn) [0x00008] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Sync.cs:174 
  at OpenRA.DefaultInputHandler.OnMouseInput (MouseInput input) [0x00020] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Input/InputHandler.cs:50 
  at OpenRA.Platforms.Default.Sdl2Input.PumpInput (IInputHandler inputHandler) [0x001e7] in /home/matthias/Entwicklung/OpenRA/OpenRA.Platforms.Default/Sdl2Input.cs:107 
  at OpenRA.Platforms.Default.Sdl2GraphicsDevice.PumpInput (IInputHandler inputHandler) [0x0000e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs:365 
  at OpenRA.Renderer.EndFrame (IInputHandler inputHandler) [0x0000e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Renderer.cs:170 
  at OpenRA.Game.RenderTick () [0x00142] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:586 
  at OpenRA.Game.Loop () [0x0013e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:696 
  at OpenRA.Game.Run () [0x00042] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Game.cs:714 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:118 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Support/Program.cs:40 
```

when you don't define sources for automatic installation:

```yaml
ModContent:
	InstallPromptMessage: Red Alert 2 requires artwork and audio from the original game.
	HeaderMessage: At the moment only a manual installation is supported.\n\nPlease refer to https://github.com/OpenRA/ra2/wiki#content
	Packages:
		base: Base Game Files
			TestFiles: ^Content/ra2/ra2.mix, ^Content/ra2/language.mix, ^Content/ra2/multi.mix
			Required: true
		music: Base Game Music
			TestFiles: ^Content/ra2/theme.mix
```

**Comment by pchote** [comment]
@Mailaender: That happened because you didn't define any discs.  I fixed your crash, and also left you a present in https://github.com/OpenRA/ra2/pull/233.

The new mscab parser means that @obrakmann should probably recheck at least that code.

File: mods/d2k/mod.yaml [hunk 4]
```diff
@@ -215,3 +199,1986 @@ GameSpeeds:
...
+						Length: 20567
+					^Content/d2k/SOUND.RS:
+						Offset: 20559431
+						Length: 1929247
+		d2k-linux: Dune 2000 (English)
+			IDFiles:
+				music/ambush.aud: bd5926a56a83bc0e49f96037e1f909014ac0772a
+				setup/setup.z: 937f5ceb1236bf3f3d4e43929305ffe5004078e7
+			Install:
+				copy: MOVIES
```
Comment (Mailaender) [inline mods/d2k/mod.yaml]: It is lower case `movies` here for whatever reason.

```
Copying /run/media/matthias/DUNE2000/MOVIES/a_br01_e.vqa -> /home/matthias/.openra/Content/d2k/Movies/A_BR01_E.VQA
System.IO.FileNotFoundException: /run/media/matthias/DUNE2000/MOVIES/a_br01_e.vqa does not exist
File name: '/run/media/matthias/DUNE2000/MOVIES/a_br01_e.vqa'
  at System.IO.File.Copy (System.String sourceFileName, System.String destFileName, Boolean overwrite) [0x001bd] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.IO/File.cs:112 
  at System.IO.File.Copy (System.String sourceFileName, System.String destFileName) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.4.0/mcs/class/corlib/System.IO/File.cs:72 
  at OpenRA.Mods.Common.Widgets.Logic.InstallFromDiscLogic+<InstallFromDisc>c__AnonStorey5.<>m__1 () <0x40fea6a0 + 0x003ef> in <filename unknown>:0 
Deleting /home/matthias/.openra/Content/d2k/Movies/A_BR01_E.VQA
```

image

File: mods/d2k/mod.yaml [hunk 5]
```diff
@@ -215,3 +199,1986 @@ GameSpeeds:
...
+					^Content/d2k/Movies/O_FINL_E.VQA: o_finl_e.vqa
+					^Content/d2k/Movies/O_LOSE_E.VQA: o_lose_e.vqa
+					^Content/d2k/Movies/O_MNTG_E.VQA: o_mntg_e.vqa
+					^Content/d2k/Movies/G_INT1_E.VQA: g_int1_e.vqa
+					^Content/d2k/Movies/G_INT2_E.VQA: g_int2_e.vqa
+					^Content/d2k/Movies/G_MAPS_E.VQA: g_maps_e.vqa
+					^Content/d2k/Movies/G_PLN2_E.VQA: g_pln2_e.vqa
+					^Content/d2k/Movies/G_PLNT_E.VQA: g_plnt_e.vqa
+					^Content/d2k/Movies/T_TITL_E.VQA: t_titl_e.vqa
+				copy: MUSIC
```
Comment (Mailaender) [inline mods/d2k/mod.yaml]: Lower case `music`.

**Comment by Mailaender** [comment]
Works :white_check_mark: 

image

**Comment by Mailaender** [comment]
Might be worth putting those lengthy definitions in their own `content.yaml` and just reference a list of MiniYaml definitions like we do for most of the rest.

**Comment by pchote** [comment]
> Might be worth putting those lengthy definitions in their own content.yaml and just reference a list of MiniYaml definitions like we do for most of the rest.

I agree in principle, but in practice this is tricky because this metadata is used by the modchooser, and loading cross-mod assets is a bit tricky.

**Comment by pchote** [comment]
Fixed d2k linux metadata.  Can you please retest?

**Comment by Mailaender** [comment]
Confirmed :white_check_mark: Dune 2000 CD installation is fixed now.

**Comment by obrakmann** [comment]
Still looks good to me. Can't really test the ms-cab extraction, though, the only .cabs I found here are Installshield ones (with the ISc* format header)

Still, :+1:

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/51bce65fee22e55479ccedfcf0124d6cf68d0d39)

**Comment by Mailaender** [comment]
> We lose TFD support until somebody who has access to the disk can set up the new metadata for it

I generated

https://gist.github.com/Mailaender/69c2410b0ee87ab50513f0dfc73692a0

using

```csharp
#region Copyright & License Information
/*
 * Copyright 2007-2016 The OpenRA Developers (see AUTHORS)
 * This file is part of OpenRA, which is free software. It is made
 * available to you under the terms of the GNU General Public License
 * as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version. For more
 * information, see COPYING.
 */
#endregion

using System;
using System.IO;
using OpenRA.FileSystem;

namespace OpenRA.Mods.Common.UtilityCommands
{
	class ListInstallShieldCabContentsCommand : IUtilityCommand
	{
		public string Name { get { return "--list-hdr"; } }

		public bool ValidateArguments(string[] args)
		{
			return args.Length == 2;
		}

		[Desc("DATA.HDR", "Lists the filenames contained within an InstallShield HDR index file")]
		public void Run(ModData modData, string[] args)
		{
			var filename = Path.GetFileName(args[1]);
			var path = Path.GetDirectoryName(args[1]);

			var fs = new OpenRA.FileSystem.FileSystem();
			fs.Mount(path, "parent");
			var package = new InstallShieldCABExtractor(fs, "parent|" + filename);

			foreach (var file in package.FileDescriptors)
			{
				Console.WriteLine("{0}:", file.Value.Filename);
				Console.WriteLine("\tOffset: {0}", file.Value.DataOffset);
				Console.WriteLine("\tLength: {0}", file.Value.NameOffset);
			}
		}
	}
}
```

**Comment by MortovMolotov** [comment]
Hey guys,

I own a TFD disc; how can I help? What do you mean by metadata in this context?

**Comment by Mailaender** [comment]
You can help test https://github.com/OpenRA/OpenRA/pull/11536.

**Comment by Stealthx3** [comment]
Is this a firstrun-only prompt like the current one? It would greatly benefit from main menu/settings menu access in the event the user would prefer to add the extra content later.

**Comment by GraionDilach** [comment]
No, it isn't. The Manage Content is always available next to the Play button in the mod selection menu.

**Comment by jtyle6** [comment]
Example:
screen shot 2016-08-13 at 7 25 52 pm

**Comment by Stealthx3** [comment]
I don't see it, if you are talking about the "Asset browser" (which would be an appropriate place to put it) or the fact that you can install music specifically from the music menu under extras in-game, then both are missing full functionality (the ability to add movies too) and the first-run import/asset grabber does not pull the videos.

screenshot from 2016-08-15 16-02-51

**Comment by GraionDilach** [comment]
There was no playtest or release since this got merged.

**Comment by Stealthx3** [comment]
Oh ok, that's a relief. The fact that imports can only happen on the first run of any mod rubbed me the wrong way a bit lol.
